### ISSUE TITLE
Account chooser dialog problem with some Android versions

### ISSUE THREAD

**Issue by eneiluj** [nextcloud/Android-SingleSignOn#60]
Hi, first of all, thanks a lot for this library, it's used in PhoneTrack-Android. Luv'it.

So far, everything works fine with Android 4.1, 5.0, 8.0, 8.1 and 9.0.

I'm testing with freshly installed Nexus 4 virtual phones without google services but have confirmation it works fine on my Nexus 5x (LineageOS 15, Android 8.1).

I'm having problems with Android 6, 7.0 and 7.1 just after choosing an account (`AccountImporter.pickNewAccount()`)

With v0.3.0 It was apparently making Nextcloud Files app crash. Here are the corresponding logs: 
```
03-30 01:46:33.888  2104  2104 E AndroidRuntime: FATAL EXCEPTION: main
03-30 01:46:33.888  2104  2104 E AndroidRuntime: Process: com.nextcloud.client, PID: 2104
03-30 01:46:33.888  2104  2104 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.nextcloud.client/com.owncloud.android.ui.activity.SsoGrantPermissionActivity}: java.lang.NullPointerException: Attempt to read from field 'java.lang.String android.accounts.Account.name' on a null object reference
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2665)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2726)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.-wrap12(ActivityThread.java)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1477)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.os.Handler.dispatchMessage(Handler.java:102)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.os.Looper.loop(Looper.java:154)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.main(ActivityThread.java:6119)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at java.lang.reflect.Method.invoke(Native Method)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
03-30 01:46:33.888  2104  2104 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to read from field 'java.lang.String android.accounts.Account.name' on a null object reference
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at com.owncloud.android.ui.activity.SsoGrantPermissionActivity.onCreate(SsoGrantPermissionActivity.java:106)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.Activity.performCreate(Activity.java:6679)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2618)
03-30 01:46:33.888  2104  2104 E AndroidRuntime:    ... 9 more
03-30 01:46:33.891  1673  1803 W ActivityManager:   Force finishing activity com.nextcloud.client/com.owncloud.android.ui.activity.SsoGrantPermissionActivity
03-30 01:46:33.914  1673  1803 W ActivityManager:   Force finishing activity net.eneiluj.nextcloud.phonetrack.dev/net.eneiluj.nextcloud.phonetrack.android.activity.SettingsActivity
```

Hanging around in the issues, I discovered I could include `-SNAPSHOT` version instead of v0.3.0. Problem is still there in a less radical form :wink: : I get an alert dialog telling me `You have to accept the requested permissions to use the single sign-on feature.` Nothing in the logs. Is it possible I did something wrong? It seems to be related to `SsoGrantPermissionActivity`. Any idea?

Ow BTW, same thing is happening with News app, there's a crash. I guess it's using SSO lib v0.3.0. Same symptoms as PhoneTrack.

**Comment by David-Development** [comment]
@eneiluj Sorry for the delay. Thanks for using the library! Glad you like it! :)

I think I fixed the crash by adding a null check (https://github.com/nextcloud/Android-SingleSignOn/blob/master/src/main/java/com/nextcloud/android/sso/AccountImporter.java#L267).
If the account is `null`, the exception (that you encountered) will be thrown. This should only happen when the app is reinstalled and there are some issues with the reauthentication. 

I'm not exactly sure what's happening with the account on your device.. It looks like the call to [`AccountImporter.getAccountForName(...)`](https://github.com/nextcloud/Android-SingleSignOn/blob/master/src/main/java/com/nextcloud/android/sso/AccountImporter.java#L266) doesn't return the account itself. Which means that something went wrong while granting the permissions of the account to your app.. (android account permissions itself). 

Basically the call to [`AccountManager.newChooseAccountIntent(...)`](https://github.com/nextcloud/Android-SingleSignOn/blob/6c6f8f6f9d43514664c04e448315560a31d7a180/src/main/java/com/nextcloud/android/sso/AccountImporter.java#L72) should be sufficient. Not sure why it's not working.. 😕(Ref: https://stackoverflow.com/a/49829475/11405479)

**Comment by tobiasKaminsky** [comment]
Do you have a branch/tag we can check out, to try it ourselves?

**Comment by eneiluj** [comment]
@David-Development The crash was just happening with v0.3.0, not with "-SNAPSHOT" version. Yes it seems the problem is not entirely related to SSO but with system permission granting system. It's happening on all devices with Android >= 6 and <= 7.1

@tobiasKaminsky Yep:
```
git clone --recurse-submodules https://gitlab.com/eneiluj/phonetrack-android -b sso-issue
```
This is an Android Studio project.

Thank you both for your answer!

**Comment by tobiasKaminsky** [comment]
@eneiluj I tried 
- your sso-issue branch with 0.3.0 instead of -SNAPSHOT
- Android 7.0
- latest Files app
--> works without problems

**Comment by eneiluj** [comment]
With a virtual device running Android AOSP (without google services), Nextcloud files app: v3.6.0 and PhoneTrack using SSO lib SNAPSHOT, I reach the account selection dialog:
<img src="https://user-images.githubusercontent.com/11291457/57310135-8e622b80-70e9-11e9-886c-e4f70ae08ba6.png" width="200" />
and after having selected one, I directly get this dialog:
<img src="https://user-images.githubusercontent.com/11291457/57310142-91f5b280-70e9-11e9-9e73-6d80115a6e9d.png" width="200" />

with SSO lib v0.3.0, I get a crash from Nextcloud files app:
<img src="https://user-images.githubusercontent.com/11291457/57311025-29a7d080-70eb-11e9-8774-2f46adca4dc3.png" width="200" />

Quite strange we get different behaviours... Are you using stable F-Droid release of Nextcloud Files app v3.6.0?

**Comment by tobiasKaminsky** [comment]
So I used
- Android 7
- Android Files app master / 3.6.0 with Android Studio
- phonetrack with master-SNAPSHOT for SSO
--> working

- Android 7
- Android Files app 3.6.0 from F-Droid
- phonetrack with master-SNAPSHOT for SSO
--> NOT working

For some reason account is null, but
https://github.com/nextcloud/Android-SingleSignOn/blob/cf6c448851a472c3ece3a4a1c42dd1d0dc7edf98/src/main/java/com/nextcloud/android/sso/AccountImporter.java#L269-L278

if account is null NextcloudFilesAppAccountPermissionNotGrantedException should be thrown…

**Comment by Valdnet** [comment]
I confirm the problem with SSO on Android 6. There is a crash of Nextcloud Android v3.6.0 after selecting an account.

App:
Nextcloud Android v3.6.0 gplay
PhoneTrack Dev  v0.0.10 with SSO

**Comment by David-Development** [comment]
@Valdnet @tobiasKaminsky @eneiluj Thanks for confirming this issue and running some tests. I managed to reproduce the issue on my android emulator (android 7). I'm working on a fix right now. It looks like you'll have to grant the Android `GET_ACCOUNTS` permission manually before being able to access the account data.. Very annoying...

**Comment by David-Development** [comment]
@tobiasKaminsky It's working if you compile it using android studio as they have the same signature then and therefore android grants you the permission to access the accounts..

**Comment by eneiluj** [comment]
@David-Development Thanks a lot for fixing this!

For the moment SSO lib acts like 'get accounts' permission is not granted with Nextcloud Files installed from F-Droid and PhoneTrack compiled with Android Studio even though I can see the permission for PhoneTrack in system settings... I get the error dialog saying "...you need to grant permission to access your accounts".

Can I assume it might work with both PhoneTrack and Files installed from F-Droid?

**Comment by David-Development** [comment]
@eneiluj Which version of the sso library are you using? Version 0.3.1 should fix this issue. However please not that you might need to add some additional code for handling the permission callbacks. 

As far as I know it should not matter which combination you use. @tobiasKaminsky Or is there a difference wether you install the files app from the play store or from the f-droid store?

**Comment by tobiasKaminsky** [comment]
> As far as I know it should not matter which combination you use. @tobiasKaminsky Or is there a difference wether you install the files app from the play store or from the f-droid store?

It should not, as the package name is the same.

**Comment by eneiluj** [comment]
@David-Development It works like a charm with v0.3.1.Thanks a lot. I can confirm `GET_ACCOUNTS` permission is required with Android 7.

**Comment by Valdnet** [comment]
@David-Development Works correctly on Android 4.4.2 and 6 :+1:

**Comment by SpinningK** [comment]
I still find this problem on a customized android 11 ROM, tried to reinstall both nextcloud and third-party app with giving all permission before logging in, but still failed.


![Screenshot_2023-02-23-13-13-21-257_News](https://user-images.githubusercontent.com/31563522/220826513-25f8b108-29f7-4677-bc35-2bf5afb413ba.png)

![Screenshot_2023-02-23-13-13-10-317_Settings](https://user-images.githubusercontent.com/31563522/220826499-b36ab73e-bd5d-4f04-81a7-0abf5bc7258f.png)

**Comment by stefan-niedermann** [comment]
@SpinningK did you also add

```xml
<queries>
    <package android:name="com.nextcloud.client" />
    <package android:name="com.nextcloud.android.beta" />
</queries>
```

to your `AndroidManifest.xml` as described in the FAQ? If not please open a new issue and provide a **minimal** sample repository with an app to reproduce the issue.
